cpu/lpc2387/periph/i2c.c

```diff
@@ -0,0 +1,479 @@
+/*
+ * Copyright (C) 2020 Beuth Hochschule für Technik Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_lpc2387
+ * @ingroup     drivers_periph_i2c
+ * @{
+ *
+ * @file
+ * @brief       Low-level I2C driver implementation for lpc23xx
+ *
+ * @author      Benjamin Valentin <benpicco@beuth-hochschule.de>
+ *
+ * @}
+ */
+
+#include <assert.h>
+#include <stdint.h>
+#include <errno.h>
+
+#include "cpu.h"
+#include "board.h"
+#include "byteorder.h"
+#include "periph_conf.h"
+#include "periph/i2c.h"
+
+#include "sched.h"
+#include "thread.h"
+#include "mutex.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#if I2C_NUMOF > 0
+static void I2C0_IRQHandler(void) __attribute__((interrupt("IRQ")));
+#endif
+#if I2C_NUMOF > 1
+static void I2C1_IRQHandler(void) __attribute__((interrupt("IRQ")));
+#endif
+#if I2C_NUMOF > 2
+static void I2C2_IRQHandler(void) __attribute__((interrupt("IRQ")));
+#endif
+
+/**
+ * We only ever need two buffers for read/write reg
+ */
+#define TRX_BUFS_MAX    (2)
+static struct i2c_ctx {
+    mutex_t lock;
+    mutex_t tx_done;
+    uint8_t *buf[TRX_BUFS_MAX];
+    uint8_t *buf_end[TRX_BUFS_MAX];
+    uint8_t *cur;
+    uint8_t *end;
+    int res;
+    uint8_t addr[TRX_BUFS_MAX];
+    uint8_t buf_num;
+    uint8_t buf_cur;
+} ctx[I2C_NUMOF];
```

Comment:
How about

```C
static struct i2c_ctx {
/* ..., same as above */
} ctx[I2C_NUMOF] = {
    { .lock = MUTEX_INIT, .tx_done = MUTEX_INIT },
    { .lock = MUTEX_INIT, .tx_done = MUTEX_INIT }
};
```

That would get rid of the call to `mutex_init()`. If you take the suggestion, please squash that right in. Otherwise, just squash. I'll do one round of testing and then finally this should get in.

Sorry for the long stall. I have a lot to do recently...

Reply:
This needed some additional `#ifdefs` that made is less pretty, but still saves about 200 bytes of ROM.

Reply:
Sounds like a good trade off :-) Please squash!

Reply:
Did you drop this by accident, or do you favor getting rid of the `#ifdef`s over the saved ROM size?

Reply:
I think I simply forgot about it - added it now.
With `USEMODULE=i2c_scan make -C examples/default` the difference is only 48 bytes, but it's something :)

And it makes `i2c_init()` simpler, which should ease adaption to #13421